include/proteus/JitEngineDevice.hpp

```diff
@@ -168,19 +168,23 @@ template <typename ImplT> class JitEngineDevice : public JitEngine {
 
     BinaryInfo &BinInfo = KernelInfo.getBinaryInfo();
 
-    if (BinInfo.hasModule()) {
-      KernelInfo.setModule(BinInfo.getModule());
-      return KernelInfo.getModule();
+    if (! BinInfo.hasModule()) {
+      std::unique_ptr<Module> ExtractedModule =
+          static_cast<ImplT &>(*this).extractModule(BinInfo);
+
+      pruneIR(*ExtractedModule);
+      runCleanupPassPipeline(*ExtractedModule);
+
+      BinInfo.setModule(std::move(ExtractedModule));
     }
 
-    std::unique_ptr<Module> ExtractedModule =
-        static_cast<ImplT &>(*this).extractModule(BinInfo);
+    std::unique_ptr<Module> KernelModule = llvm::CloneModule(BinInfo.getModule());
 
-    pruneIR(*ExtractedModule);
-    runCleanupPassPipeline(*ExtractedModule);
+    pruneKernelIR(*KernelModule, KernelInfo.getName());
+    pruneIR(*KernelModule);
```

Comment:
I'm not sure we need to run `pruneIR` again on the `KernelModule` since this pruning must have been applied to the base `ExtractedModule`. Please check.